### ISSUE TITLE
Need to agree on competition rules

### ISSUE THREAD

**Issue by raaperrotta** [raaperrotta/nothanks#2]
What number of players should compete?
How many games?
What pot size?
Should strategies be able to appear more than once in a game (and have to play themselves)?
How do we crown a victor?

**Comment by raaperrotta** [comment]
In this first take at `compete.py`, the competition is built from 10,000 3-player games, 10,000 4-player games, and 10,000 5-player games. In each game, players ante a combined 60 chips (60 because it is nicely divisible by 2, 3, 4, and 5) and winners split the pot. For example, in a 4 player game each player pays 15 to play; in the event of a single winner, the winner takes all 60; in the case of a two-way tie the pot is split 30, 30.

Each game involves a random selection of the competing strategies, chosen without replacement (so no player object plays itself). This means that not all players play the same number of games.

**Comment by ginop** [comment]
How many chips are used in a regular game (the paper version)? 4 and 5 players seems reasonable, although we could standardize it to just 4 or just 5 if we wanted. Also, I like the idea of strategies playing against themselves as well (since the inputs will be different for each it won't just be a trivial result). Can we select random players *with* replacements instead of without?

Tracking scores should be easy. The ante system you described reduces to the same net-zero scoring I had in mind. Since there are no physical chips involved, it is just as easy for all players to ante a combined 1 chip (each gives 1/n for n players). Either way, the final score is a combination of win rate and total games played, so should be normalized by total games to extract the percent won.

**Comment by raaperrotta** [comment]
Each player gets 11 chips to start. The game rules allow for 3 to 5 players, but I am happy to just pick a number of players for now.

I agree, playing against similar strategies would be good. Our current implementation makes this tricky; we only create one instance of each class. Then we let that instance play many games. We could create fresh instances for each game but then players would not be able to "learn" or adapt their play to the opponents. What do you think about that? I don't have any plans to make a learning algorithm but you mentioned it at one point and maybe someday I would want to. Do you want an instance to play multiple games in a competition or does that not matter?

We could sample without replacement from a list including duplicate instances of each strategy class. Then we could score the strategy as the average of all instances using that strategy.

The easy to-do's:

- [x] change ante from 60 combined to 1 combined (1/num_players each)
- [x] normalize competition result by games played

**Comment by ginop** [comment]
I think each game can be a new instance. Training should take place outside
of the competition, not during it. That was statistics can be tracked for
many games with identical participants.

**Comment by raaperrotta** [comment]
- [x] move strategy instantiation inside competition loop.
- [x] select strategies from list of names WITH replacement

**Comment by raaperrotta** [comment]
How many games should be played in a competition?
Do we care that each strategy may not play the same number of games?
Should the normalization reflect that?
Do we want to play 3, 4, and 5 player games and combine the results or just pick a number of players?

**Comment by ginop** [comment]
Let's do 3, 4, and 5 and combine the results. The impact of each should be weighted by number of players to reflect the relative chances of winning based on random selection. (We should verify later that a game of all players choosing randomly returns zero average score.)

It is okay that players will play different numbers of games, as long as the difference is small compared to the mean. Out of 10k, I think a few games off will not matter. The normalization should still reflect the exact number played.

**Comment by raaperrotta** [comment]
Okay, I divided by the total number of games, but it sounds like we should track the number of games played by each strategy (how do you count games where the strategy is represented multiple times?) and divide by that. Or something like that.

Do you want to implement that (however you see fit)?

**Comment by ginop** [comment]
Yes, I will work on that! A strategy that participates in the same game
from multiple seats should keep the stats as separate games. In the end, I
would like to look at stats about wins against particular opponents, wins
with certain scores, seat arrangements, who knows what. So can we log all
of the games' results for future analysis? Or perhaps make that an output
option if it is too slow to run every time?

**Comment by raaperrotta** [comment]
I doubt that would be slow at all. I think that’s a great idea. If you take
a crack at that you might consider using nested dictionaries. They convert
to pandas tables nicely. And they are fast and readable (named indexing
versus numeric).
